File: python/three_sum.py
from typing import List

def three_sum(nums: List[int]) -> List[List[int]]:
    nums.sort()
    n = len(nums)
    result = []

    for i in range(n - 2):
        # 跳过重复的 i
        if i > 0 and nums[i] == nums[i - 1]:
            continue
        # 剪枝：当前最小和 > 0，后面更大，退出
        if nums[i] + nums[i + 1] + nums[i + 2] > 0:
            break
        # 剪枝：当前数 + 最大两个 < 0，跳过
        if nums[i] + nums[n - 2] + nums[n - 1] < 0:
            continue

        left, right = i + 1, n - 1
        while left < right:
            total = nums[i] + nums[left] + nums[right]
            if total == 0:
                result.append([nums[i], nums[left], nums[right]])
                left += 1
                right -= 1
                # 跳过重复
                while left < right and nums[left] == nums[left - 1]:
                    left += 1
                while left < right and nums[right] == nums[right + 1]:
                    right -= 1
            elif total < 0:
                left += 1
            else:
                right -= 1

    return result


# 测试
print(three_sum([-1, 0, 1, 2, -1, -4]))  # [[-1, -1, 2], [-1, 0, 1]]
print(three_sum([0, 1, 1]))              # []
print(three_sum([0, 0, 0]))              # [[0, 0, 0]]

File: TS/three_sum.ts
function threeSum(nums: number[]): number[][] {
    nums.sort((a, b) => a - b);
    const n = nums.length;
    const result: number[][] = [];

    for (let i = 0; i < n - 2; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;
        if (nums[i] + nums[i + 1] + nums[i + 2] > 0) break;
        if (nums[i] + nums[n - 2] + nums[n - 1] < 0) continue;

        let left = i + 1;
        let right = n - 1;

        while (left < right) {
            const total = nums[i] + nums[left] + nums[right];
            if (total === 0) {
                result.push([nums[i], nums[left], nums[right]]);
                left++;
                right--;
                while (left < right && nums[left] === nums[left - 1]) left++;
                while (left < right && nums[right] === nums[right + 1]) right--;
            } else if (total < 0) {
                left++;
            } else {
                right--;
            }
        }
    }

    return result;
}

// 测试
console.log(threeSum([-1, 0, 1, 2, -1, -4]));  // [[-1, -1, 2], [-1, 0, 1]]
console.log(threeSum([0, 1, 1]));               // []
console.log(threeSum([0, 0, 0]));               // [[0, 0, 0]]
